src/test/SubscribableRequestQueue.test.ts

```diff
@@ -0,0 +1,262 @@
+      expect(id1).to.not.equal(id2);
+    });
+  });
+
+  describe("addRequestToQueue", () => {
+    it("queues a request", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+      const promise = queue.addRequestToQueue("test", id, () => delay(TIMEOUT, "foo"));
+      expect(queue.hasRequest("test")).to.be.true;
+      const result = await promise;
+      expect(result).to.equal("foo");
+    });
+
+    it("does not queue a request if the subscriber id is invalid", () => {
+      const queue = new SubscribableRequestQueue();
+      expect(() => queue.addRequestToQueue("test", -1, () => delay(TIMEOUT, "foo"))).to.throw();
+      expect(() => queue.addRequestToQueue("test", 0, () => delay(TIMEOUT, "foo"))).to.throw();
+    });
+
+    it("does not queue a request if the subscriber has been deleted", () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+      queue.removeSubscriber(id);
+      expect(() => queue.addRequestToQueue("test", id, () => delay(TIMEOUT, "foo"))).to.throw();
+    });
+
+    it("creates unique promises without duplicating work when two subscribers queue the same key", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id1 = queue.addSubscriber();
+      const id2 = queue.addSubscriber();
+
+      const promise1 = queue.addRequestToQueue("test", id1, () => delay(TIMEOUT, "foo"));
+      const promise2 = queue.addRequestToQueue("test", id2, () => delay(TIMEOUT, "bar"));
+      expect(queue.hasRequest("test")).to.be.true;
+      expect(promise1).to.not.equal(promise2);
+
+      const [result1, result2] = await Promise.all([promise1, promise2]);
+      expect(result1).to.equal("foo");
+      expect(result2).to.equal("foo");
+    });
+
+    it("creates different requests when one subscriber queues two keys", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+
+      const promise1 = queue.addRequestToQueue("test1", id, () => delay(TIMEOUT, "foo"));
+      const promise2 = queue.addRequestToQueue("test2", id, () => delay(TIMEOUT, "bar"));
+      expect(queue.hasRequest("test1")).to.be.true;
+      expect(queue.hasRequest("test2")).to.be.true;
+      expect(promise1).to.not.equal(promise2);
+
+      const [result1, result2] = await Promise.all([promise1, promise2]);
+      expect(result1).to.equal("foo");
+      expect(result2).to.equal("bar");
+    });
+
+    it("rejects and reissues a request when one subscriber queues the same key twice", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+
+      const promise1 = queue.addRequestToQueue("test", id, () => delay(TIMEOUT, "foo"));
+      const promise2 = queue.addRequestToQueue("test", id, () => delay(TIMEOUT, "bar"));
+      expect(queue.hasRequest("test")).to.be.true;
+      expect(await isRejected(promise1)).to.be.true;
```

Comment:
Would it be a plausible alternative to just drop the reference to the "foo" result and simply replace it with the "bar" result, without rejecting anything?